b2g/components/RemoteControlService.jsm

```diff
@@ -0,0 +1,384 @@
+    certService.getOrCreateCert("RemoteControlService", {
+      handleCert: function(cert, result) {
+        if(result) {
+          aReject("getOrCreateCert " + result);
+          return;
+        } else {
+          try {
+            // Try to get random port
+            let ios = Cc["@mozilla.org/network/io-service;1"].getService(Ci.nsIIOService);
+            let socket;
+            for (let i = 100; i; i--) {
+              let temp = Cc["@mozilla.org/network/tls-server-socket;1"].createInstance(Ci.nsITLSServerSocket);
+              temp.init(self._port, false, MAX_CLIENT_CONNECTIONS);
+              temp.serverCert = cert;
+
+              let allowed = ios.allowPort(temp.port, "tls");
+              if (!allowed) {
+                DEBUG && debug("Warning: obtained TLSServerSocket listens on a blocked port: " + temp.port);
+              }
+
+              if (!allowed && self._port == -1) {
+                DEBUG && debug("Throw away TLSServerSocket with bad port.");
+                temp.close();
+                continue;
+              }
+
+              socket = temp;
+              break;
+            }
+
+            if (!socket) {
+              throw new Error("No socket server available. Are there no available ports?");
+            }
+
+            DEBUG && debug("Listen on port " + socket.port + ", " + MAX_CLIENT_CONNECTIONS + " pending connections");
+
+            socket.serverCert = cert;
+            socket.setSessionCache(false);
+            socket.setSessionTickets(false);
+            socket.setRequestClientCertificate(Ci.nsITLSServerSocket.REQUEST_NEVER);
+
+            socket.asyncListen(self);
+            self._port = socket.port;
+            self._serverSocket = socket;
+          } catch (e) {
+            DEBUG && debug("Could not start server on port " + self._port + ": " + e);
+            aReject("Start TLSSocketServer fail");
+            return;
+          }
+
+          // Register mDNS remote control service with this._port
+          if (("@mozilla.org/toolkit/components/mdnsresponder/dns-sd;1" in Cc)) {
+            let serviceInfo = Cc["@mozilla.org/toolkit/components/mdnsresponder/dns-info;1"]
+                                .createInstance(Ci.nsIDNSServiceInfo);
+            serviceInfo.serviceType = REMOTECONTROL_PREF_MDNS_SERVICE_TYPE;
+            serviceInfo.serviceName = Services.prefs.getCharPref(REMOTECONTROL_PREF_MDNS_SERVICE_NAME);
+            serviceInfo.port = self._port;
+
+            let mdns = Cc["@mozilla.org/toolkit/components/mdnsresponder/dns-sd;1"]
+                         .getService(Ci.nsIDNSServiceDiscovery);
+            self._mDNSRegistrationHandle = mdns.registerService(serviceInfo, null);
+          }
+
+          aResolve();
+          self._serverStatus = SERVER_STATUS.STARTED;
+        }
+      }
+    });
+  },
+
+  // Notifies this server that the given connection has been closed.
+  _connectionClosed: function(connection) {
```

Comment:
I still prefer only passing connectionId instead of passing the connection object.